### DISCUSSION TITLE
Recipes for RAG prompts

### DISCUSSION THREAD

**Discussion by matiasmolinas** [microsoft/AI-Chat-App-Hack#3] [Ideas]
I think a key part in this kind of solutions is how to make good prompts. 
What do you think?
At the end of the day, this is how to "program" with natural language the interaction with the llm. And how to ask something from the llm, and because you have freedom to write it you can do something very useful or not.
As an exercise I am working on create an extension of the codelab https://github.com/Azure-Samples/azure-search-openai-demo and I found this is one of the most challenging parts.
What you think is a good recipe? Some tips that works for me are, use always a chat with at least 3 different steps (that is guide and explain to the llm what you want to do), set a system message for each chat, describe how to do the task, use examples.
Any other tip?
What are your thoughts about that?

**Comment by pamelafox** [comment]
That is an awesome question! My colleague @charris-msft wrote up a doc with his process on coming up with a new prompt. Chris, perhaps you can share it here? He also has some recommended videos that he learned from.

**Reply by pamelafox** [reply]
Here's the video that Chris liked:

https://www.youtube.com/watch?v=jC4v5AS4RIM


And then he analyzed and designed new prompts based on criteria of task, context, exemplar, persona, format, and tone.

My experience from running https://github.com/Azure-Samples/ai-rag-chat-evaluator on various prompts is that the GPT models are generally very helpful and inclined towards question answering, so prompts should be about getting responses in specific format / adhering to specific sources. ie. don't worry about trying to make the LLM helpful, that's the default.

**Reply by matiasmolinas** [reply]
Amazing! thanks for the link to the video and your comments. I am going to run some experiments using the ai-rag-chat-evaluator.

**Comment by charris-msft** [comment]
Thanks @pamelafox!

@matiasmolinas - I would definitely encourage you to check out the video. The document I wrote summarizes the important points I took from Jeff Su's video, how I used them to modify my prompt, and the results I received, but Jeff's is the source material and will provide helpful context that is not included in my doc.

If you watch his video and would like to see my example, let me know and I will share the original prompt and the variations I tried.

**Reply by matiasmolinas** [reply]
Hi, yes, definitely. I am going to watch the video and make some experiments and soon I going to share my comments. And please yes, if possible share your experiment. I will be great to analyze it and try to make additional variations. Thanks a lot!

**Reply by charris-msft** [reply]
My goal was very specific. To increase trust in the AI Chat responses, I wanted to ensure each response included citations that clearly supported it.

For the original prompt and each variation below, I tested combinations of the app configuration settings to ensure I was getting accurate citations regardless of the configuration. These settings included:
* Search Results (int - 3 or 5)
* Semantic Ranker (bool)
* Retrieval Mode (Hybrid, Vectors, Text)

Legend based on Jeff Su's video:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/74415662/01f5a827-37af-47d0-b23e-ced2aacbd3e1)

Original prompt:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/74415662/8baf5a97-d7c7-4cdd-b2fc-779d73a2f1b6)

Variation #1:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/74415662/1f022f43-93b2-47a0-a06c-4872c0a9537e)

Variation #2:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/74415662/123bc222-0a78-456d-ada8-87884a3795f3)

Variation #3:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/74415662/1715ddf6-ca58-4e13-a7e2-14663299f42c)

In my very limited testing, I found that the 4th prompt did the best.
I also found that regardless of the prompt the citations were always inaccurate with Retrieval Mode set to "Vectors" as opposed to Hybrid or Text.

I hope this helps and I'd love to hear your results!

**Reply by matiasmolinas** [reply]
This is amazing. What about use chat? I am already using chat with mainly applying the similar criteria  (not all of them, I found your examples very useful, and based on that I am trying to improve my "recipes"). But I found in the past that is a plus organize the same components of the dialog that you use in a chat. e.g. starting with system and putting there the persona and the intial description of the task, or that the persona is expert on that task, describe the task as a user, interact with the assistant about examples. etc. Is it also a valid approach, do you think a chat is more valuable than only one prompt? In my experiments I achieve  better results using chats.. but may be is because how I am writing the prompts..

**Reply by charris-msft** [reply]
I was just asking a single question, but the app uses the chat api. I should have clarified that the prompt I was working with is the system prompt rather than the user prompt. In my case the user prompt was “does my plan cover annual eye exams?”.
